integration-core/src/main/java/org/symphonyoss/integration/core/service/CryptoServiceImpl.java

```diff
@@ -126,99 +94,89 @@ public String encrypt(String plainText, String key) throws CryptoException {
-    try {
-      char[] keyBytes = key.toCharArray();
-      SecretKeyFactory factory = SecretKeyFactory.getInstance(KEY_ALGORITHM);
-      PBEKeySpec spec = new PBEKeySpec(keyBytes, salt, ITERATIONS, KEY_SIZE);
-      SecretKey secretKey = factory.generateSecret(spec);
-      return new SecretKeySpec(secretKey.getEncoded(), SPEC_ALGORITHM);
-    } catch (NoSuchAlgorithmException e) {
-      throw new CryptoException(logMessage.getMessage(NO_SUCH_ALGORITHM), e,
-          logMessage.getMessage(NO_SUCH_ALGORITHM_SOLUTION, KEY_ALGORITHM));
-    } catch (InvalidKeySpecException e) {
-      throw new CryptoException(logMessage.getMessage(INVALID_KEY_SPEC), e,
-          logMessage.getMessage(INVALID_KEY_SPEC_SOLUTION, "PBEKeySpec"));
-    }
-  }
-
-  // The following 3 methods were created to decrease the amount of duplicate code (exception
-  // handling), since those exceptions are not throwable from the Unit Test and our minimum
-  // coverage is 90% per project.
-  private Cipher getCipher() throws CryptoException {
-    try {
-      return Cipher.getInstance(TRANSFORMATION);
-    } catch (NoSuchAlgorithmException e) {
-      throw new CryptoException(logMessage.getMessage(NO_SUCH_ALGORITHM), e,
-          logMessage.getMessage(NO_SUCH_ALGORITHM_SOLUTION, SPEC_ALGORITHM));
-    } catch (NoSuchPaddingException e) {
-      throw new CryptoException(logMessage.getMessage(NO_SUCH_PADDING), e,
-          logMessage.getMessage(NO_SUCH_PADDING_SOLUTION, PADDING));
-    }
-  }
-
-  private void initCipher(Cipher cipher, int mode, Key key, AlgorithmParameterSpec spec)
+  /**
+   * This method was created to decrease the amount of duplicate code (exception
+   * handling), since those exceptions are not throwable from the Unit Test and our minimum
+   * coverage is 90% per project.
+   *
+   * It performs an encryption or decryption according to the passed MODE.
+   *
+   * @param key Key used to derive a PasswordBasedKey used during the cryptography process.
+   * @param text Text to encrypted or decrypted.
+   * @param salt Salt used to reinforce the security.
+   * @param encrypt MODE, set to TRUE to encrypt or FALSE to decrypt.
+   * @return Encrypted or decrypted text as an array of bytes.
+   * @throws CryptoException In case of error during the cryptography process.
+   */
+  private byte[] doCipher(String key, String text, byte[] salt, boolean encrypt)
       throws CryptoException {
     try {
-      if (spec == null) {
-        cipher.init(mode, key);
-      } else {
-        cipher.init(mode, key, spec);
+      if (encrypt) {
+        // Derive a Password-based key from the key parameter + salt
+        byte[] pbKey = PBKDF.PBKDF2_SHA256(key.getBytes(CHARSET), salt, NUMBER_OF_ITERATIONS);
+        // Encrypt the plain text
+        KeyIdentifier keyId = new KeyIdentifier(new byte[STREAM_ID_SIZE], 0L, 0L);
+        IClientCryptoHandler clientCryptoHandler = new ClientCryptoHandler();
```

Comment:
You can create this object as a class attribute. I believe this approach will help you to mock this class on unit tests and to check the exception flows.

Reply:
The implementation has changed.